graph/network/diffusion_test.go

```diff
@@ -0,0 +1,495 @@
+		if !mat.EqualApprox(l, test.want, tol) {
+			t.Errorf("unexpected result for test %d:\ngot:\n% .2v\nwant:\n% .2v",
+				i, mat.Formatted(l), mat.Formatted(test.want))
+		}
+	}
+}
+
+type sortedNodeGraph struct {
+	graph.Graph
+}
+
+func (g sortedNodeGraph) Nodes() []graph.Node {
+	n := g.Graph.Nodes()
+	sort.Sort(ordered.ByID(n))
+	return n
+}
+
+var diffuseToEquilibriumTests = []struct {
+	g       []set
+	builder builder
+	h       map[int64]float64
+	damp    float64
+	tol     float64
+	iter    int
+
+	want   map[int64]float64
+	wantOK bool
+}{
+	{
+		g:       grid(5),
+		builder: simple.NewUndirectedGraph(),
+		h:       map[int64]float64{0: 1},
+		damp:    0.85,
+		tol:     1e-6,
+		iter:    1e4,
+
+		want: map[int64]float64{
+			A: 0.025000, B: 0.037500, C: 0.037500, D: 0.037500, E: 0.025000,
+			F: 0.037500, G: 0.050000, H: 0.050000, I: 0.050000, J: 0.037500,
+			K: 0.037500, L: 0.050000, M: 0.050000, N: 0.050000, O: 0.037500,
+			P: 0.037500, Q: 0.050000, R: 0.050000, S: 0.050000, T: 0.037500,
+			U: 0.025000, V: 0.037500, W: 0.037500, X: 0.037500, Y: 0.025000,
+		},
+		wantOK: true,
+	},
+	{
+		// Example graph from http://en.wikipedia.org/wiki/File:PageRanks-Example.svg 16:17, 8 July 2009
+		g: []set{
+			A: nil,
+			B: linksTo(C),
+			C: linksTo(B),
+			D: linksTo(A, B),
+			E: linksTo(D, B, F),
+			F: linksTo(B, E),
+			G: linksTo(B, E),
+			H: linksTo(B, E),
+			I: linksTo(B, E),
+			J: linksTo(E),
+			K: linksTo(E),
+		},
+		builder: simple.NewDirectedGraph(),
+		h: map[int64]float64{
+			A: 1. / 11.,
+			B: 1. / 11.,
+			C: 1. / 11.,
+			D: 1. / 11.,
+			E: 1. / 11.,
+			F: 1. / 11.,
+			G: 1. / 11.,
+			H: 1. / 11.,
+			I: 1. / 11.,
+			J: 1. / 11.,
+			K: 1. / 11.,
+		},
+		damp: 0.85,
+		tol:  1e-6,
+		iter: 1e4,
+
+		// This does not look like Page Rank because we do not
+		// do the random node hops. An alternative Laplacian
+		// value that does do that would replicate PageRank. This
+		// is left as an excercise for the reader.
+		want: map[int64]float64{
+			A: 0.227273,
+			B: 0.386364,
+			C: 0.386364,
+			D: 0.000000,
+			E: 0.000000,
+			F: 0.000000,
+			G: 0.000000,
+			H: 0.000000,
+			I: 0.000000,
+			J: 0.000000,
+			K: 0.000000,
+		},
+		wantOK: true,
+	},
+	{
+		g: []set{
+			A: linksTo(B, C),
+			B: linksTo(D, C),
+			C: nil,
+			D: nil,
+			E: linksTo(F),
+			F: nil,
+		},
+		builder: simple.NewDirectedGraph(),
+		h:       map[int64]float64{A: 1, E: -10},
+		tol:     1e-6,
+		iter:    3,
+
+		want: map[int64]float64{
+			A: 0, B: 0, C: 0.75, D: 0.25, E: 0, F: -10,
+		},
+		wantOK: true,
+	},
+	{
+		g: []set{
+			A: linksTo(B, C),
+			B: linksTo(D, C),
+			C: nil,
+			D: nil,
+			E: linksTo(F),
+			F: nil,
+		},
+		builder: simple.NewUndirectedGraph(),
+		h:       map[int64]float64{A: 1, E: -10},
+		damp:    0.85,
+		tol:     1e-6,
+		iter:    1e4,
+
+		want: map[int64]float64{
+			A: 0.25, B: 0.375, C: 0.25, D: 0.125, E: -5, F: -5,
+		},
+		wantOK: true,
+	},
```

Comment:
An additional test with `wantOK: false` could be:
```
g: []set{A: linksTo(B), B: linksTo(C)},
h: map[int64]float64{B: 1},
```
It should oscillate between `{0, 1, 0}` and `{0.5, 0, 0.5}`.

Reply:
I've added that test here since it shows that a fix to the vector data return loop needed to happen.

PTAL